sdkbuild/java.go

```diff
@@ -0,0 +1,209 @@
+	// caches and do extra validation, the build won't be used by NewCommand.
+	Build bool
+	// If present, this directory is expected to exist beneath base dir. Otherwise
+	// a temporary dir is created.
+	DirName string
+	// If present, applied to build commands before run. May be called multiple
+	// times for a single build.
+	ApplyToCommand func(context.Context, *exec.Cmd) error
+}
+
+// JavaProgram is a Java-specific implementation of Program.
+type JavaProgram struct {
+	dir string
+}
+
+var _ Program = (*JavaProgram)(nil)
+
+// BuildJavaProgram builds a Java program. If completed successfully, this can
+// be stored and re-obtained via JavaProgramFromDir() with the Dir() value (but
+// the entire BaseDir must be present too).
+func BuildJavaProgram(ctx context.Context, options BuildJavaProgramOptions) (*JavaProgram, error) {
+	if options.BaseDir == "" {
+		return nil, fmt.Errorf("base dir required")
+	} else if _, err := os.Stat(filepath.Join(options.BaseDir, "build.gradle")); err != nil {
+		return nil, fmt.Errorf("failed finding build.gradle in base dir: %w", err)
+	} else if options.HarnessDependency == "" {
+		return nil, fmt.Errorf("harness dependency required")
+	} else if options.MainClass == "" {
+		return nil, fmt.Errorf("main class required")
+	}
+
+	// Create temp dir if needed that we will remove if creating is unsuccessful
+	success := false
+	var dir string
+	if options.DirName != "" {
+		dir = filepath.Join(options.BaseDir, options.DirName)
+	} else {
+		var err error
+		dir, err = os.MkdirTemp(options.BaseDir, "program-")
+		if err != nil {
+			return nil, fmt.Errorf("failed making temp dir: %w", err)
+		}
+		defer func() {
+			if !success {
+				// Intentionally swallow error
+				_ = os.RemoveAll(dir)
+			}
+		}()
+	}
+	j := &JavaProgram{dir}
+
+	// If we depend on SDK via path, built it and get the JAR
+	isPathDep := strings.ContainsAny(options.Version, `/\`)
+	if isPathDep {
+		cmd := j.buildGradleCommand(ctx, options.Version, false, options.ApplyToCommand, "jar", "gatherRuntimeDeps")
+		if err := cmd.Run(); err != nil {
+			return nil, fmt.Errorf("failed building Java SDK: %w", err)
+		}
+		// Copy JARs to sdkjars dir
+		sdkJarsDir := filepath.Join(dir, "sdkjars")
+		if err := copy.Copy(filepath.Join(options.Version, "temporal-sdk/build/libs"), sdkJarsDir); err != nil {
+			return nil, fmt.Errorf("failed copying lib JARs: %w", err)
+		}
+		if err := copy.Copy(filepath.Join(options.Version, "temporal-sdk/build/runtimeDeps"), sdkJarsDir); err != nil {
+			return nil, fmt.Errorf("failed copying runtime JARs: %w", err)
+		}
+	}
+
+	// Create build.gradle and settings.gradle
+	temporalSDKDependency := ""
+	if isPathDep {
+		temporalSDKDependency = "implementation fileTree(dir: 'sdkjars', include: ['*.jar'])"
+	} else if options.Version != "" {
+		temporalSDKDependency = fmt.Sprintf("implementation 'io.temporal:temporal-sdk:%v'",
+			strings.TrimPrefix(options.Version, "v"))
+	}
+	buildGradle := `
```

Comment:
is `sdkbuild` is intended to be a separate library I am not sure we can hardcode all these config files

Reply:
we allow the `go.mod` file to be customized for the builder for Go

Reply:
> is sdkbuild is intended to be a separate library 

Yes

> I am not sure we can hardcode all these config files

For most, we always have, this refactor just moves it

> we allow the go.mod file to be customized for the builder for Go

Not following what you mean here

Reply:
When this wasn't a library it didn't matter that it was hard coded, but as a library I think it does matter. You allow the `go.mod` to be customized, but not the other config files like the gradle. I am arguing all languages build config files should be customizable.

Reply:
I only allow what needs to be customized to support my two needs for flexibility reasons. I would take pieces of the go.mod if that were reasonable. This library should not be considered to have a stable API. I may need to update the docs with that. This is a simple code sharing exercise between two testing tools that don't require backwards compatibility. Don't think of it as an officially distributed Go library. It won't even have a version.

I fully expect as more flexibility is needed it is added. I tried to make the most shared I could without spending a lot oftime here (which means things like package.json and pyproject.toml have a lot shared, but go.mod has project name and such that made it make less sense).